lib/dry/result_matcher/evaluator.rb

```diff
@@ -0,0 +1,40 @@
+module Dry
+  module ResultMatcher
+    class Evaluator
+      attr_reader :result
+      attr_reader :output
+
+      def initialize(result, cases)
+        @result = result
+        @output = nil
+        @matched = false
+
+        cases.each do |name, kase|
+          define_singleton_method(name) do |pattern = nil, &block|
+            handle_case(kase, pattern, &block)
+          end
+        end
+      end
+
+      def call(&block)
+        block.call(self)
+        output
+      end
+
+      private
+
+      def handle_case(kase, pattern, &block)
+        return(output) if matched?
+
+        if kase.matches?(pattern, result)
+          @matched = true
+          @output = block.call(kase.resolve(result))
+        end
```

Comment:
Sorry for pushing you like that but does it really have to mutate? Are you being concerned about performance? I typically try to avoid mutating objects and it's possible in 99.9% of the cases w/o negative impact on the performance :)